### PR TITLE
Add `sorted_ipc_queues` invariant

### PR DESCRIPTION
**Pull request by michaelmcinerney**
Please see the commit messages for what is here. It might be best to review the files in the order in which they would appear in jedit. So for example, AInvs before Refine, and earlier files in AInvs before later files in AInvs.

As usual, there are many lemmas that we may want to move up, or to have in `master`, and I would be happy to do that. I have moved many of the lemmas around, and I'll try to indicate them with comments. The diff is not very accurate, and some of what appears as green or red isn't anything I have actually touched.

I'll try some things out to hopefully reduce the number of times I remove `comp_apply` from the simp set.

I have weakened `valid_ep'` and `valid_ntfn'` by removing some of the properties that can be crossed from the abstract state analogues. This was done because proving distinctness was a bit of a pain with the ordered insert. So I have added some asserts in the Haskell for distinctness of these queues.

### CODE REVIEW

File: proof/invariant-abstract/DetSchedInvs_AI.thy
Comment (michaelmcinerney) [inline proof/invariant-abstract/DetSchedInvs_AI.thy:3051]: I did decide in AInvs to remove `opt_map_Some_comp` from the simp set, but it ended up back in the simp set during Refine (in a way I haven't tracked down yet). I think it would be best to not have it in the simp set at all.

File: proof/invariant-abstract/IpcDet_AI.thy
Comment (michaelmcinerney) [inline proof/invariant-abstract/IpcDet_AI.thy:124]: Everything between here and `opt_ord_R` is old and was moved from `DetSchedSchedule`

File: proof/invariant-abstract/IpcDet_AI.thy
Comment (michaelmcinerney) [inline proof/invariant-abstract/IpcDet_AI.thy:266]: This is a generalisation of the older `opt_ord`. I added the `_R` to the name and allowed it to take an arbitrary relation. Then I define `opt_ord` below as it was before

File: proof/invariant-abstract/IpcDet_AI.thy
Comment (michaelmcinerney) [inline proof/invariant-abstract/IpcDet_AI.thy:307]: Lemmas from here until `total_opt_ord` are old and were just moved

File: lib/Lib.thy
Comment (Xaphiosis) [inline lib/Lib.thy:2272]: This is a pretty generic name to go into lib; this looks something like a `in_set_zipD` based on the names of lemmas above

File: spec/haskell/src/SEL4/Object/TCB.lhs
Comment (Xaphiosis) [inline spec/haskell/src/SEL4/Object/TCB.lhs:1122]: this looks much nicer than the tcbEPFindIndex way

File: spec/abstract/IpcCancel_A.thy
Comment (Xaphiosis) [inline spec/abstract/IpcCancel_A.thy:315]: comprehension nitpick: if it's placing the tptr into a list such that it's sorted, does it still make sense to call it "append" without comment?

File: proof/invariant-abstract/ARM/ArchDetSchedAux_AI.thy
Comment (Xaphiosis) [inline proof/invariant-abstract/ARM/ArchDetSchedAux_AI.thy:85]: if you had a lifting lemma for valid_sched_pred_strong, you could apply that here and then wp most of the conditions away; the hoare_lift_Pf way looks a bit uphill and happens 3 times in this file, so presumably also on RISCV64

File: proof/invariant-abstract/DetSchedInvs_AI.thy
Comment (Xaphiosis) [reply proof/invariant-abstract/DetSchedInvs_AI.thy:3051]: That looks painful. I'm not really an expert on the construction and deployment of these, so I'll leave it to @lsf37. I can say that it should be possible to track down where it gets reintroduced in Refine and kick it out early. Maybe in Invariants_H?

File: spec/abstract/IpcCancel_A.thy
Comment (lsf37) [reply spec/abstract/IpcCancel_A.thy:315]: Good point, it'd be more like an `insert`. I suspect the name hasn't changed in C either, but we could still do better in the specs.

File: lib/Lib.thy
Comment (lsf37) [reply lib/Lib.thy:2272]: +1 for `in_set_zipD`

File: proof/crefine/RISCV64/IpcCancel_C.thy
Comment (lsf37) [inline proof/crefine/RISCV64/IpcCancel_C.thy:2557]: These were all there before, but if we're touching it, let's remove the extra parentheses:

```suggestion
  "\<lbrakk> valid_ep' ep s; isSendEP ep \<or> isRecvEP ep \<rbrakk>
   \<Longrightarrow> epQueue ep \<noteq> [] \<and> (\<forall>t\<in>set (epQueue ep). tcb_at' t s)"
```

File: proof/crefine/RISCV64/SR_lemmas_C.thy
Comment (lsf37) [inline proof/crefine/RISCV64/SR_lemmas_C.thy:1215]: (also old, but nicer this way)

```suggestion
  \<Longrightarrow> ntfnQueue (ntfnObj ntfn) \<noteq> [] \<and> (\<forall>t\<in>set (ntfnQueue (ntfnObj ntfn)). tcb_at' t s)"
```

File: proof/invariant-abstract/Tcb_AI.thy
Comment (Xaphiosis) [inline proof/invariant-abstract/Tcb_AI.thy:1744]: This looks a bit like monadic_rewrite_bind immediately applies and would yield precisely the kind of goals which you wrote as assumptions. Bit confused how this is working.

File: proof/refine/RISCV64/CNodeInv_R.thy
Comment (Xaphiosis) [inline proof/refine/RISCV64/CNodeInv_R.thy:6947]: since you're in a `proof` block, you can use `note` to mark these as `[simp del]` for the entire lemma, which will save the repetition

File: proof/invariant-abstract/ARM/ArchDetSchedAux_AI.thy
Comment (lsf37) [inline proof/invariant-abstract/ARM/ArchDetSchedAux_AI.thy:41]: We should be proving `P (tcbs_of s)` here (and wherever we can), because that's the more general case that includes `prios_of` and all other tcb projections.

File: proof/invariant-abstract/ARM/ArchDetSchedAux_AI.thy
Comment (lsf37) [inline proof/invariant-abstract/ARM/ArchDetSchedAux_AI.thy:94]: Since these 3 always occur together: I wonder if there is a combinator that we could build so that we only have to prove one lemma and then apply a rule or something else that is automatic to get all three together. Not for this PR, but in general for projections.

Comment (lsf37) [reply proof/invariant-abstract/ARM/ArchDetSchedAux_AI.thy:94]: (No action necessary for this, but if someone has an idea, do let me know)

File: proof/invariant-abstract/DetSchedAux_AI.thy
Comment (Xaphiosis) [inline proof/invariant-abstract/DetSchedAux_AI.thy:47]: oo, another candidate for that lifting rule

File: proof/invariant-abstract/DetSchedAux_AI.thy
Comment (Xaphiosis) [inline proof/invariant-abstract/DetSchedAux_AI.thy:105]: yes, lifting rule

File: proof/invariant-abstract/DetSchedAux_AI.thy
Comment (Xaphiosis) [inline proof/invariant-abstract/DetSchedAux_AI.thy:737]: vertical compression: `fastforce+` or auto or something

File: proof/invariant-abstract/DetSchedAux_AI.thy
Comment (Xaphiosis) [inline proof/invariant-abstract/DetSchedAux_AI.thy:972]: nice that you isolated this, thanks

File: proof/invariant-abstract/IpcDet_AI.thy
Comment (Xaphiosis) [inline proof/invariant-abstract/IpcDet_AI.thy:119]: yes, these could go into `lib`, suggest `_nth` instead of `'`

File: proof/invariant-abstract/IpcDet_AI.thy
Comment (Xaphiosis) [reply proof/invariant-abstract/IpcDet_AI.thy:266]: bikeshed: opt_ord_R vs opt_ord_rel? we don't seem to name defs with stuff like `R` as far as I can tell

File: proof/invariant-abstract/IpcDet_AI.thy
Comment (Xaphiosis) [inline proof/invariant-abstract/IpcDet_AI.thy:272]: this doesn't make sense ... if you have a definition where you don't even bother with a type signature, and then you immediately add that definition to `[simp]`, how is it any better than an abbreviation?

File: proof/invariant-abstract/IpcDet_AI.thy
Comment (Xaphiosis) [inline proof/invariant-abstract/IpcDet_AI.thy:358]: hmm, yeah my brain would prefer `opt_ord_rel R` to `opt_ord_R R`

File: proof/invariant-abstract/IpcDet_AI.thy
Comment (Xaphiosis) [inline proof/invariant-abstract/IpcDet_AI.thy:403]: another lib candidate; if you're ever in the mood for janitorial work, there's a fair few things I've seen in this review we could move to lib

File: proof/invariant-abstract/DetSchedSchedule_AI.thy
Comment (Xaphiosis) [inline proof/invariant-abstract/DetSchedSchedule_AI.thy:1090]: anothere one for team lift

File: proof/invariant-abstract/DetSchedSchedule_AI.thy
Comment (Xaphiosis) [inline proof/invariant-abstract/DetSchedSchedule_AI.thy:27304]: I would like to take this opportunity to express concern at the size of DetSchedSchedule_AI on rt. My browser took longer than my tea does to brew in order to open this file's diff for review.

**Comment by Xaphiosis** [comment]
I like how you structured the proof of this invariant, and the way the asserts are lining up and not polluting Refine works out well. I didn't spot anything style-wise, and also appreciated where you pointed out a bunch of lemmas were moved.

I think the outstanding non-minor issues from my side are:
- the over-use of `hoare_lift_Pf` rules indicates missing lifting rules; `hoare_lift_Pf` rules are generally avoided if possible because, amongst other issues, they nearly always apply and so can loop, and are the antithesis of automation
- you are correct about the `opt_map_Some_comp` situation. Apparently, this might be able to be addressed with making `||>` a definition, and we might want to give this a go before we merge this PR

File: lib/Lib.thy
Comment (michaelmcinerney) [reply lib/Lib.thy:2272]: This is an old lemma that I moved to `Lib` because it appears in a few places and if I hadn't been moved to `Lib`, I would need to state it twice, due to the import structure not being linear. So I could rename it and then do one separate commit to use the new name in the proofs, and then fix up the name in the main commit

File: proof/invariant-abstract/IpcDet_AI.thy
Comment (michaelmcinerney) [reply proof/invariant-abstract/IpcDet_AI.thy:266]: I think I'll go with `opt_ord_rel`

File: proof/invariant-abstract/Tcb_AI.thy
Comment (michaelmcinerney) [reply proof/invariant-abstract/Tcb_AI.thy:1744]: I've removed this `monadic_rewrite` lemma and instead used `monadic_rewrite_refine_valid` and `monadic_rewrite_bind_tail` within the proof where the previous lemma was used.

File: proof/invariant-abstract/IpcDet_AI.thy
Comment (michaelmcinerney) [reply proof/invariant-abstract/IpcDet_AI.thy:272]: Good point. I've made this an abbreviation

File: lib/Lib.thy
Comment (michaelmcinerney) [reply lib/Lib.thy:2272]: And this lemma with the name `set_zip_helper` occurs on `master`, too. Considering that we want `Lib` to be as similar as possible between `master` and `rt`, I suppose I'll have to make a PR for `master` moving this lemma and renaming it...

File: proof/invariant-abstract/IpcDet_AI.thy
Comment (michaelmcinerney) [reply proof/invariant-abstract/IpcDet_AI.thy]: Done

**Comment by michaelmcinerney** [comment]
I think I have addressed most of what we would like to change for this PR, with the exception of renaming `tcb_ep_append/tcbEPAppend` to `tcb_ep_insert/tcbEPInsert`. There are also further lemmas that I could move to `Lib`. If we would very much like those tasks done as part of this PR, then I could do that, but I would prefer doing those in separate PRs, if that is OK.

File: lib/Lib.thy
Comment (Xaphiosis) [reply lib/Lib.thy]: Good idea. It doesn't look big, so it'll go quickly.

File: proof/invariant-abstract/Tcb_AI.thy
Comment (Xaphiosis) [reply proof/invariant-abstract/Tcb_AI.thy]: Cool, makes sense now

**Comment by Xaphiosis** [comment]
> I think I have addressed most of what we would like to change for this PR, with the exception of renaming `tcb_ep_append/tcbEPAppend` to `tcb_ep_insert/tcbEPInsert`. There are also further lemmas that I could move to `Lib`. If we would very much like those tasks done as part of this PR, then I could do that, but I would prefer doing those in separate PRs, if that is OK.

Fine to do separately. Don't forget to make the issue for the misc/lifting/hoare_liftPf situation if you haven't already. Someone should double-check the squashed commits once they're squashed, but from my side this looks good.

**Review by Xaphiosis — APPROVED** [review]

**Comment by michaelmcinerney** [comment]
Here's my issue for the heap projections: https://github.com/seL4/l4v/issues/841

File: spec/haskell/src/SEL4/Object/TCB.lhs
Comment (lsf37) [reply spec/haskell/src/SEL4/Object/TCB.lhs:1122]: Agreed.

**Review by lsf37 — APPROVED** [review]
Good work! I have nothing further to add over the minor queries from before and Raf's thorough review.

**Comment by lsf37** [comment]
And nice job on resolving the `comp_apply` issue, that went even better than I expected.

**Comment by michaelmcinerney** [comment]
I've force pushed to update this PR so that it now consists of the commits from https://github.com/seL4/l4v/pull/845 ("Some changes for `Lib`") cherry-picked onto the `rt` branch, plus squashed commits for the new work for `sorted_ipc_queues`

**Pull request merged**
